volatility3/framework/objects/utility.py

```diff
@@ -115,50 +119,65 @@ def gather_contiguous_bytes_from_address(layer, address: int, count: int) -> byt
 
     data = b""
 
-    left_to_read = count
-
-    # read as many pages as possible that are contiguous
-    # if the first page is missed, we re-raise the InvalidAddressException
-    # if we have at least 1 page that was read succesfully,
-    # then we try to construct a string from it
-    while left_to_read > 0:
-        # compute aligned address of current page and next the page
-        aligned = address & ~0xFFF
-        next_page = aligned + 0xFFF + 1
-
-        # all fits on the current page, last read
-        if address + left_to_read < next_page:
-            try:
-                data += layer.read(address, left_to_read)
-            except exceptions.InvalidAddressException:
-                # if we have data, just break the loop
-                if data:
-                    break
-                # Raise if no data was read as this means the first page was invalid
-                else:
-                    raise
-
-            left_to_read = 0
-
-        else:
-            # how many bytes are left on the current page
-            len_to_read = next_page - address
-
-            try:
-                data += layer.read(address, len_to_read)
-            except exceptions.InvalidAddressException:
-                if data:
-                    break
-                # Raise if no data was read as this means the first page was invalid
-                else:
-                    raise
-
-            address += len_to_read
-            left_to_read -= len_to_read
+    last_address = None
+
+    for address, length, _, _, _ in data_layer.mapping(
+        offset=starting_address, length=count, ignore_errors=True
+    ):
+        # Used to track when we hit a paged out page
+        if not last_address:
+            last_address = address + length
```

Comment:
After the first time round the loop, `last_address` will have been set, so this won't get updated and will stay the same for every iteration of the loop after.  I think last_address needs to be update after the data addition has been done...

Reply:
Yea, its there now.. not sure how it got missed

Reply:
Still doesn't look like it gets incremented except if it didn't exist previously?

Reply:
overhauled